cmd/metal-api/internal/datastore/machine.go

```diff
@@ -487,3 +492,121 @@ func (rs *RethinkStore) FindWaitingMachine(partitionid, sizeid string) (*metal.M
 	}
 	return &newMachine, nil
 }
+
+func electMachine(allMachines, projectMachines metal.Machines, tags []string) metal.Machine {
+	rackids := make([]string, 0)
+
+	for _, m := range allMachines {
+		rackids = append(rackids, m.RackID)
+	}
+
+	allRacks := groupByRack(allMachines, rackids)
+	projectRacks := groupByRack(projectMachines, rackids)
+	projectWinners := electRacks(projectRacks)
+	taggedMachines := flatten(filter(groupByTags(projectMachines), tags...))
+	taggedRacks := groupByRack(taggedMachines, rackids)
+	tagWinners := electRacks(taggedRacks)
+	winners := intersect(tagWinners, projectWinners)
+
+	candidates := allMachines
+	if c := flatten(filter(allRacks, winners...)); len(c) > 0 {
+		candidates = c
+	} else if c := flatten(filter(allRacks, tagWinners...)); len(c) > 0 {
+		candidates = c
+	} else if c := flatten(filter(allRacks, projectWinners...)); len(c) > 0 {
+		candidates = c
+	}
```

Comment:
This looks like a perfect example for using a functional implementation like: https://github.com/samber/lo

But i am unsure if this is worth adding another dependency.

Reply:
Instead of using generic functions we added this to a specific type. Overhead is actually pretty small so we decided it's ok and did not introduce another library.